components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -1122,240 +1339,125 @@ void FeedV2Builder::NotifyUpdateCompleted() {
   std::tie(hash_, subscribed_count_) =
       GetFeedHashAndSubscribedCount(channels, publishers, feed_etags_);
 
-  // Fire all the pending callbacks.
-  for (auto& callback : current_update_->callbacks) {
-    std::move(callback).Run();
-  }
-
   // If we have a |next_update_| then request an update with that data.
+  auto callbacks = std::move(current_update_->callbacks);
   current_update_ = std::move(next_update_);
   next_update_ = std::nullopt;
 
-  // Notify listeners of updated hash.
-  for (const auto& listener : listeners_) {
-    listener->OnUpdateAvailable(hash_);
-  }
-
-  if (current_update_) {
-    UpdateData(current_update_->settings);
-  }
-}
-
-void FeedV2Builder::GenerateFeed(
-    UpdateSettings settings,
-    mojom::FeedV2TypePtr type,
-    base::OnceCallback<mojom::FeedV2Ptr()> build_feed,
-    BuildFeedCallback callback) {
-  UpdateData(
-      std::move(settings),
+  // Fire all the pending callbacks. Wait for them all to complete before we
+  // fire off the next update.
+  auto on_notify_completed = base::BarrierClosure(
```

Comment:
Looks a bit weired to me..
1. The method named `NotifyUpdateCompleted` but the actual notifications seems to be postoned to `on_notify_completed`.
2. We modify some class field like an update is finished but in fact we are still waiting to the callbacks.

Reply:
Ah nice catch - we don't actually need this anymore. It was to make sure the data wasn't changed while we were running callbacks.

However, because we take a copy of the data now we don't need this.

Also noticed a bug where |next_updates_| wouldn't get triggered :fearful: which I fixed.